src/components/gltf-model-plus.js

```diff
@@ -341,127 +340,221 @@ function runMigration(version, json) {
   }
 }
 
-const loadLightmap = async (parser, materialIndex) => {
-  const lightmapDef = parser.json.materials[materialIndex].extensions.MOZ_lightmap;
-  const [material, lightMap] = await Promise.all([
-    parser.getDependency("material", materialIndex),
-    parser.getDependency("texture", lightmapDef.index)
-  ]);
-  material.lightMap = lightMap;
-  material.lightMapIntensity = lightmapDef.intensity !== undefined ? lightmapDef.intensity : 1;
-  return lightMap;
-};
-
 let ktxLoader;
 
-export async function loadGLTF(src, contentType, onProgress, jsonPreprocessor) {
-  let gltfUrl = src;
-  let fileMap;
-
-  if (contentType && (contentType.includes("model/gltf+zip") || contentType.includes("application/x-zip-compressed"))) {
-    fileMap = await extractZipFile(gltfUrl);
-    gltfUrl = fileMap["scene.gtlf"];
+class GLTFHubsPlugin {
+  constructor(parser, jsonPreprocessor, fileMap) {
+    this.parser = parser;
+    this.jsonPreprocessor = jsonPreprocessor;
+    this.fileMap = fileMap;
+
+    // The latest GLTFLoader doesn't use ImageBitmapLoader for Firefox
+    // because the latest ImageBitmapLoader passes an option parameter
+    // to createImageBitmap() but Firefox createImageBitmap() fails
+    // if an option parameter is passed (known bug and already reported to bugzilla).
+    // But our r111 based ImageBitmapLoader doesn't pass the option parameter yet
+    // so we can use ImageBitmapLoader even for Firefox.
+    // When we replace our Three.js fork with the latest official one
+    // we need to revisit this workaround.
+    if (!parser.textureLoader.isImageBitmapLoader && typeof createImageBitmap !== undefined) {
+      parser.textureLoader = new THREE.ImageBitmapLoader(parser.options.manager);
+    }
   }
 
-  const loadingManager = new THREE.LoadingManager();
-  loadingManager.setURLModifier(getCustomGLTFParserURLResolver(gltfUrl));
-  const gltfLoader = new THREE.GLTFLoader(loadingManager);
+  beforeRoot() {
+    const parser = this.parser;
+    const jsonPreprocessor = this.jsonPreprocessor;
 
-  // TODO some models are loaded before the renderer exists. This is likely things like the camera tool and loading cube.
-  // They don't currently use KTX textures but if they did this would be an issue. Fixing this is hard but is part of
-  // "taking control of the render loop" which is something we want to tackle for many reasons.
-  if (!ktxLoader && AFRAME && AFRAME.scenes && AFRAME.scenes[0]) {
-    ktxLoader = new KTX2Loader(loadingManager).detectSupport(AFRAME.scenes[0].renderer);
-  }
+    //
+    if (jsonPreprocessor) {
+      parser.json = jsonPreprocessor(parser.json);
+    }
 
-  if (ktxLoader) {
-    gltfLoader.setKTX2Loader(ktxLoader);
-  }
+    //
+    let version = 0;
+    if (
```

Comment:
Not particularly important but eventually we might want this to be its own MozHubsComponetnsExtension that just handles the components extension and not the other weird hubs specific workarounds. Doesn't really matter too much now since the components stuff is all still very tightly integrated into aframe and the rest of Hubs.

Reply:
Added a comment so far and would like to think of the cleaning up later.